lib/ProductOpener/PackagingStats.pm

```diff
@@ -0,0 +1,647 @@
+
+Compute stats (means etc.) for packaging materials, aggregated at the countries, categories and materials levels
+
+=cut
+
+sub compute_stats_for_all_materials ($packagings_materials_stats_ref, $delete_values = 1) {
+
+	foreach my $country_ref (values %{$packagings_materials_stats_ref->{countries}}) {
+		foreach my $category_ref (values %{$country_ref->{categories}}) {
+			foreach my $material_ref (values %{$category_ref->{materials}}) {
+				foreach my $field ("products_percent", "weight_100g", "weight_100g_material") {
+					if (defined $material_ref->{$field}) {
+						# Compute stats
+						compute_stats_for_values($material_ref->{$field});
+						# Delete individual values
+						if ($delete_values) {
+							delete $material_ref->{$field}{values};
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return;
+}
+
+=head2 generate_packaging_stats_for_query($name, $query_ref)
+
+Generate packaging stats for products matching a specific query.
+
+Stats are saved in .json format in $data_root/data/categories_stats/
+and in JSON format in $www_root/data/categories_stats/
+
+=head3 Arguments
+
+=head4 name $name
+
+=head4 query reference $query_ref
+
+=cut
+
+sub generate_packaging_stats_for_query ($name, $query_ref, $quiet = 0) {
+
+	# we will filter out empty and obsolete products
+	$query_ref->{'empty'} = {"\$ne" => 1};
+	$query_ref->{'obsolete'} = {"\$ne" => 1};
+
+	# fields to retrieve
+	my $fields_ref = {
+		code => 1,
+		countries_tags => 1,
+		categories_tags => 1,
+		packagings => 1,
+		packagings_materials => 1,
+		product_quantity => 1,
+	};
+
+	my $socket_timeout_ms = 3 * 60 * 60 * 60000;    # 3 hours
+	my $products_collection = get_products_collection({timeout => $socket_timeout_ms});
+
+	my $products_count = $products_collection->count_documents($query_ref);
+
+	$quiet or print STDERR "$name: $products_count products\n";
+
+	my $cursor = $products_collection->query($query_ref)->sort({created_t => 1})->fields($fields_ref);
+
+	$cursor->immortal(1);
+
+	my $total = 0;
+
+	my $packagings_stats_ref
+		= {};    # Stats for shapes and materials, with input at the level of each packaging component of each product
+	my $packagings_materials_stats_ref = {};    # Stats for parent materials, with input at the level of each product
+
+	# Export packaging components of all products to a CSV file
+	my $filehandle;
+	my $filename = "$www_root/data/$name.csv";
+	open($filehandle, ">:encoding(UTF-8)", $filename)
+		or die("Could not write " . $filename . " : $!\n");
+	my $csv = Text::CSV->new(
+		{
+			eol => "\n",
+			sep => "\t",
+			quote_space => 0,
+			binary => 1
+		}
+	) or die "Cannot use CSV: " . Text::CSV->error_diag();
+
+	# Print the header line with fields names
+	$csv->print(
+		$filehandle,
+		[
+			"code", "countries_tags", "categories_tags", "number_of_units",
+			"shape", "material", "parent_material", "recycling",
+			"weight", "weight_measured", "weight_specified", "quantity_per_unit"
+		]
+	);
```

Comment:
style: typically the kind of thing I would like to have in a separate method, to avoid hindering code read. (something like: `my ($filehandle, $csv) =  init_products_packaging_components_csv()`)

Reply:
Good idea, will do.